kratos/processes/flush_streams_process.py

```diff
@@ -0,0 +1,103 @@
+                      - Finalize
+        @ingroup KratosCore
+    """
+
+    def __init__(self,
+                 _: KratosMultiphysics.Model,
+                 parameters: KratosMultiphysics.Parameters):
+        super().__init__()
+        parameters.ValidateAndAssignDefaults(self.GetDefaultParameters())
+        self.__buffer_interval: float = parameters["interval"].GetDouble()
+        self.__flush_stdout: bool = parameters["stdout"].GetBool()
+        self.__flush_stderr: bool = parameters["stderr"].GetBool()
+        self.__stages: "set[str]" = set()
+        self.__last_flush: float = time.time()
+
+        # Parse trigger stages
+        admissible_stages: "set[str]" = set(["Initialize",
+                                             "BeforeSolutionLoop",
+                                             "InitializeSolutionStep",
+                                             "FinalizeSolutionStep",
+                                             "BeforeOutputStep",
+                                             "AfterOutputStep",
+                                             "Finalize"])
+
+        for stage_parameter in parameters["stages"].values():
+            stage_name = stage_parameter.GetString()
+            if stage_name not in admissible_stages:
+                raise ValueError(f"invalid stage {stage_name}. Options: {' '.join(admissible_stages)}")
+            self.__stages.add(stage_name)
+
+    def Execute(self) -> None:
+        current_time = time.time()
+        if self.__buffer_interval <= current_time - self.__last_flush:
+            self.__Flush()
+            self.__last_flush = current_time
+
+    def ExecuteInitialize(self) -> None:
+        if "Initialize" in self.__stages:
+            self.Execute()
+
+    def ExecuteBeforeSolutionLoop(self) -> None:
+        if "BeforeSolutionLoop" in self.__stages:
+            self.Execute()
+
+    def ExecuteInitializeSolutionStep(self) -> None:
+        if "InitializeSolutionStep" in self.__stages:
+            self.Execute()
+
+    def ExecuteFinalizeSolutionStep(self) -> None:
+        if "FinalizeSolutionStep" in self.__stages:
+            self.Execute()
+
+    def ExecuteBeforeOutputStep(self) -> None:
+        if "BeforeOutputStep" in self.__stages:
+            self.Execute()
+
+    def ExecuteAfterOutputStep(self) -> None:
+        if "AfterOutputStep" in self.__stages:
+            self.Execute()
+
+    def ExecuteFinalize(self) -> None:
+        if "Finalize" in self.__stages:
+            self.Execute()
+
+    @classmethod
+    def GetDefaultParameters(cls) -> KratosMultiphysics.Parameters:
+        return KratosMultiphysics.Parameters("""{
+            "interval" : 0.0,
+            "stdout" : true,
+            "stderr" : true,
+            "stages" : ["FinalizeSolutionStep"]
+        }""")
+
+    def __Flush(self) -> None:
+        if self.__flush_stdout:
+            sys.stdout.flush()
+        if self.__flush_stderr:
+            sys.stderr.flush()
```

Comment:
In my experience also a lot of stuff gets stuck in the logger:
https://github.com/KratosMultiphysics/Kratos/blob/245da2c2b63da441df93c57071451b31172f0d60/applications/CoSimulationApplication/python_scripts/co_simulation_analysis.py#L99-L102

Reply:
wait, which stream is the logger using? I thought it was reading to stdout.

Reply:
It depends on how the Logger is constructed. It can use to have `std::out`, or `std::fstream` or other stream types as well. So using as @philbucher mentioned makes sense.

Reply:
https://github.com/KratosMultiphysics/Kratos/blob/aacb6c1c3549c2a5e4fa37b11a5970b6c8540e6a/applications/RANSApplication/python_scripts/formulations/utilities.py#L335-L346